### ISSUE TITLE
Allow option to use system fltk installation

### ISSUE THREAD

**Issue by davidovski** [pwiecz/go-fltk#117]
Since fltk is already included within this project in binary form, it is not possible to use this toolkit on systems that have not been pre-compiled for, this includes other architectures or operating systems.

I am unable to use this module as-is on a musl-based linux distribution and I expect the same will be the case on any flavour of BSD. The current linux libfltk.a is compiled against glibc, which will only work on other 'standard' linux distributions, with missing symbol errors occuring on other platforms, at best.

I am not too well-versed in go modules, so I'm not sure how this would be best implemented. I would much prefer the option to use the system's installed fltk libraries, however, to avoid having to bundle any fltk within this module. Alternatively the option to build the libraries from source when building this module would suffice also, allowing for finer control of the version of fltk that is supported. 

Having binary-only dependencies that are included directly in the project may also pose a security hazard: these binaries, if modified, may contain malicious code. The trust lies entirely upon the maintainers of this repository. (not to mention having frequently-updated binaries in a git repository greatly adds to its overall size!)

**Comment by pwiecz** [comment]
Unfortunately, afaik, only Fltk 1.3 is included blin most distributions, and this library wraps a development version of Fltk 1.4.

**Comment by davidovski** [comment]
That appears to be the case, looks like the next best option would be to build fltk 1.4 when building the package, however as far as I've seen, there is no elegant way of building non-go code within a module. Currently testing if its viable to use go:generate or something similar to achieve this, in #118, however I seem to be having trouble.

**Comment by MoAlyousef** [comment]
Can you try with:
```diff
diff --git a/fltk b/fltk
--- a/fltk
+++ b/fltk
@@ -1 +1 @@
-Subproject commit 17f61f923b758f4a54c6c9b6a5cb43cae8430159
+Subproject commit 17f61f923b758f4a54c6c9b6a5cb43cae8430159-dirty
diff --git a/fltk.go b/fltk.go
index 5ca08c7..046a40a 100644
--- a/fltk.go
+++ b/fltk.go
@@ -3,9 +3,9 @@ package fltk
 //go:generate sh fltk_build.sh
 
 /*
-#cgo CPPFLAGS: -I${SRCDIR}/include
+#cgo CPPFLAGS: -I${SRCDIR}/fltk
 #cgo CXXFLAGS: -std=c++11
-#cgo LDFLAGS: ${SRCDIR}/fltk/lib/linux/x64/libfltk.a ${SRCDIR}/fltk/lib/linux/x64/libfltk_images.a ${SRCDIR}/fltk/lib/linux/x64/libfltk_z.a ${SRCDIR}/fltk/lib/linux/x64/libfltk_jpeg.a ${SRCDIR}/fltk/lib/linux/x64/libfltk_png.a ${SRCDIR}/fltk/lib/linux/x64/libfltk_gl.a -lGLU -lGL -lXrender -lXcursor -lXfixes -lXext -lXft -lfontconfig -lXinerama -lpthread -ldl -lm -lX11 -lcairo -lpango-1.0 -lpangoxft-1.0 -lgobject-2.0 -lpangocairo-1.0
+#cgo LDFLAGS: ${SRCDIR}/fltk/lib/libfltk.a ${SRCDIR}/fltk/lib/libfltk_images.a ${SRCDIR}/fltk/lib/libfltk_z.a ${SRCDIR}/fltk/lib/libfltk_jpeg.a ${SRCDIR}/fltk/lib/libfltk_png.a ${SRCDIR}/fltk/lib/libfltk_gl.a -lGLU -lGL -lXrender -lXcursor -lXfixes -lXext -lXft -lfontconfig -lXinerama -lpthread -ldl -lm -lX11 -lcairo -lpango-1.0 -lpangoxft-1.0 -lgobject-2.0 -lpangocairo-1.0
 #include <stdlib.h>
 #include "fltk.h"
 */
diff --git a/fltk_build.sh b/fltk_build.sh
index bca7722..3cc6745 100755
--- a/fltk_build.sh
+++ b/fltk_build.sh
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-git submodules update --init --recursive
+git submodule update --init --recursive
 
 cd fltk
 
@@ -18,13 +18,20 @@ CMAKE_FLAGS="-DFLTK_BUILD_TEST=OFF
              -DFLTK_BUILD_EXAMPLES=OFF
              -DOPTION_USE_SYSTEM_LIBJPEG=OFF
              -DOPTION_USE_SYSTEM_LIBPNG=OFF
-             -DOPTION_USE_SYSTEM_ZLIB=OFF "
+             -DOPTION_USE_SYSTEM_ZLIB=OFF
+             -DOPTION_USE_WAYLAND=OFF "
 
 case "$platform" in 
     "macos")
         CMAKE_FLAGS="$CMAKE_FLAGS -DCMAKE_OSX_ARCHITECTURES=x86_64"
+        ;;
+
+    "linux")
+        CMAKE_FLAGS="$CMAKE_FLAGS -DOPTION_USE_WAYLAND=OFF -DOPTION_USE_CAIRO=ON -DOPTION_USE_PANGO=ON"
+        ;;
+
 esac
 
 cmake -B . $CMAKE_FLAGS
-cmake --build .
+cmake --build . --parallel
```

**Comment by davidovski** [comment]
Looks good, I'll try building with these and see how far I can get!

**Comment by davidovski** [comment]
Looks like im getting errors when trying to build a project that uses go-fltk with these changes, it is unable to link to the libraries in `fltk_bin`, since they do not exist. I don't think that go generate is being called when trying to build go-fltk as a module of a different project.

Trying to manually run `go generate` in the go/pkg/mod/ directory of go-fltk fails since this directory is not a git repository, so submodules cannot be fetched. Possible fix is modifying `fltk_build.sh` to fetch a distributed tarball of the source instead. 

From my quick research, I've not found any way to ensure that `go generate` is called when building a module, though there might be some workaround that I'm missing.

**Comment by MoAlyousef** [comment]
Yeah I think Go would benefit from allowing an optional build script (a la build.rs for Rust) which would also be automatically run for dependencies.

**Comment by pwiecz** [comment]
In my opinion if there is no automatic way to build a binary dependency by running `go get`, we have to provide the binaries ourselves.
User can build the binaries themself using `go generate`, and the ones provided in the repository (right now) are built using github workers using the github worklows, so user can download them directly from github action artifacts.
